# Constants & Physical Values
- **scipy.constants**: Collection of physical and mathematical constants
- **Purpose**: Provide accurate values for scientific calculations
- **Source**: CODATA (Committee on Data for Science and Technology) recommended values

Key features:
- Physical constants (c, h, G, etc.)
- Mathematical constants (pi, golden ratio)
- Unit conversion factors
- Consistent precision across all values

In [1]:
import scipy.constants as const
import numpy as np

# scipy.constants provides constants as float values
print("Module imported successfully")

Module imported successfully


## Fundamental Physical Constants

Most commonly used constants in physics and engineering:

| Constant | Variable | Unit | Description |
|----------|----------|------|-------------|
| Speed of light | `c` | m/s | Electromagnetic wave speed in vacuum |
| Planck constant | `h` | J·s | Quantum of action |
| Reduced Planck | `hbar` | J·s | h / (2π) |
| Gravitational constant | `G` | m³/(kg·s²) | Universal gravitation |
| Elementary charge | `e` | C | Electron charge |
| Boltzmann constant | `k` | J/K | Thermal energy relation |
| Avogadro constant | `N_A` | 1/mol | Particles per mole |
| Gas constant | `R` | J/(mol·K) | Ideal gas law |

In [2]:
# Speed of light
print("Speed of light (c):", const.c, "m/s")

# Planck constant
print("Planck constant (h):", const.h, "J·s")

# Reduced Planck constant (h-bar)
print("Reduced Planck (ℏ):", const.hbar, "J·s")

# Gravitational constant
print("Gravitational constant (G):", const.G, "m³/(kg·s²)")

# Elementary charge
print("Elementary charge (e):", const.e, "C")

# Boltzmann constant
print("Boltzmann constant (k):", const.k, "J/K")

Speed of light (c): 299792458.0 m/s
Planck constant (h): 6.62607015e-34 J·s
Reduced Planck (ℏ): 1.0545718176461565e-34 J·s
Gravitational constant (G): 6.6743e-11 m³/(kg·s²)
Elementary charge (e): 1.602176634e-19 C
Boltzmann constant (k): 1.380649e-23 J/K


In [3]:
# Avogadro constant
print("Avogadro constant (N_A):", const.N_A, "1/mol")

# Gas constant
print("Gas constant (R):", const.R, "J/(mol·K)")

# Stefan-Boltzmann constant
print("Stefan-Boltzmann (σ):", const.sigma, "W/(m²·K⁴)")

# Electron mass
print("Electron mass (m_e):", const.m_e, "kg")

# Proton mass
print("Proton mass (m_p):", const.m_p, "kg")

# Neutron mass
print("Neutron mass (m_n):", const.m_n, "kg")

Avogadro constant (N_A): 6.02214076e+23 1/mol
Gas constant (R): 8.31446261815324 J/(mol·K)
Stefan-Boltzmann (σ): 5.6703744191844314e-08 W/(m²·K⁴)
Electron mass (m_e): 9.1093837139e-31 kg
Proton mass (m_p): 1.67262192595e-27 kg
Neutron mass (m_n): 1.67492750056e-27 kg


## Mathematical Constants

Common mathematical constants:

| Constant | Variable | Value | Description |
|----------|----------|-------|-------------|
| Pi | `pi` | 3.14159... | Circle constant |
| Golden ratio | `golden` | 1.61803... | (1 + √5) / 2 |
| Golden ratio conjugate | `golden_ratio` | 0.61803... | (√5 - 1) / 2 |

**Note**: For most cases, use `numpy.pi` instead

In [4]:
# Pi
print("Pi (π):", const.pi)
print("NumPy Pi:", np.pi)
print("Same value:", const.pi == np.pi)

# Golden ratio
print("\nGolden ratio (φ):", const.golden)
print("Calculated: (1 + √5) / 2 =", (1 + np.sqrt(5)) / 2)

# Golden ratio conjugate
print("\nGolden ratio conjugate:", const.golden_ratio)
print("Relationship: φ - 1 =", const.golden - 1)

Pi (π): 3.141592653589793
NumPy Pi: 3.141592653589793
Same value: True

Golden ratio (φ): 1.618033988749895
Calculated: (1 + √5) / 2 = 1.618033988749895

Golden ratio conjugate: 1.618033988749895
Relationship: φ - 1 = 0.6180339887498949


## Unit Conversion Factors

### Distance Conversions
All distance units convert **to meters**:

```python
value_in_meters = value_in_other_unit * conversion_factor
```

In [5]:
# Distance conversions (to meters)
print("=== Distance Conversions ===")
print("1 inch =", const.inch, "meters")
print("1 foot =", const.foot, "meters")
print("1 yard =", const.yard, "meters")
print("1 mile =", const.mile, "meters")
print("1 nautical mile =", const.nautical_mile, "meters")
print("1 angstrom =", const.angstrom, "meters")
print("1 astronomical unit =", const.astronomical_unit, "meters")
print("1 light year =", const.light_year, "meters")
print("1 parsec =", const.parsec, "meters")

# Example: Convert 5 miles to meters
miles = 5
meters = miles * const.mile
print(f"\nExample: {miles} miles = {meters:.2f} meters")

=== Distance Conversions ===
1 inch = 0.0254 meters
1 foot = 0.30479999999999996 meters
1 yard = 0.9143999999999999 meters
1 mile = 1609.3439999999998 meters
1 nautical mile = 1852.0 meters
1 angstrom = 1e-10 meters
1 astronomical unit = 149597870700.0 meters
1 light year = 9460730472580800.0 meters
1 parsec = 3.085677581491367e+16 meters

Example: 5 miles = 8046.72 meters


### Mass Conversions
All mass units convert **to kilograms**:

```python
value_in_kg = value_in_other_unit * conversion_factor
```

In [6]:
# Mass conversions (to kilograms)
print("=== Mass Conversions ===")
print("1 gram =", const.gram, "kg")
print("1 metric ton =", const.metric_ton, "kg")
print("1 pound (lb) =", const.pound, "kg")
print("1 ounce (oz) =", const.ounce, "kg")
print("1 stone =", const.stone, "kg")
print("1 grain =", const.grain, "kg")
print("1 long ton =", const.long_ton, "kg")
print("1 short ton =", const.short_ton, "kg")
print("1 troy ounce =", const.troy_ounce, "kg")

# Example: Convert 150 pounds to kg
pounds = 150
kg = pounds * const.pound
print(f"\nExample: {pounds} pounds = {kg:.2f} kg")

=== Mass Conversions ===
1 gram = 0.001 kg
1 metric ton = 1000.0 kg
1 pound (lb) = 0.45359236999999997 kg
1 ounce (oz) = 0.028349523124999998 kg
1 stone = 6.3502931799999995 kg
1 grain = 6.479891e-05 kg
1 long ton = 1016.0469088 kg
1 short ton = 907.1847399999999 kg
1 troy ounce = 0.031103476799999998 kg

Example: 150 pounds = 68.04 kg


### Time Conversions
All time units convert **to seconds**:

```python
value_in_seconds = value_in_other_unit * conversion_factor
```

In [7]:
# Time conversions (to seconds)
print("=== Time Conversions ===")
print("1 minute =", const.minute, "seconds")
print("1 hour =", const.hour, "seconds")
print("1 day =", const.day, "seconds")
print("1 week =", const.week, "seconds")
print("1 year (Julian) =", const.year, "seconds")

# Example: Convert 2.5 hours to seconds
hours = 2.5
seconds = hours * const.hour
print(f"\nExample: {hours} hours = {seconds:.0f} seconds")

=== Time Conversions ===
1 minute = 60.0 seconds
1 hour = 3600.0 seconds
1 day = 86400.0 seconds
1 week = 604800.0 seconds
1 year (Julian) = 31536000.0 seconds

Example: 2.5 hours = 9000 seconds


### Temperature Conversions

Temperature conversions are **not simple multipliers** (except zero):

| Constant | Value | Description |
|----------|-------|-------------|
| `zero_Celsius` | 273.15 K | 0°C in Kelvin |
| `degree_Fahrenheit` | 5/9 K | Temperature difference scale |

**Conversion formulas**:
- Celsius to Kelvin: `K = C + 273.15`
- Fahrenheit to Celsius: `C = (F - 32) × 5/9`
- Fahrenheit to Kelvin: `K = (F - 32) × 5/9 + 273.15`

In [8]:
# Temperature constants
print("=== Temperature Constants ===")
print("0°C in Kelvin:", const.zero_Celsius, "K")
print("1°F difference:", const.degree_Fahrenheit, "K")

# Manual conversions (SciPy doesn't have direct functions)
def celsius_to_kelvin(celsius):
    return celsius + const.zero_Celsius

def fahrenheit_to_celsius(fahrenheit):
    return (fahrenheit - 32) * 5/9

def fahrenheit_to_kelvin(fahrenheit):
    return fahrenheit_to_celsius(fahrenheit) + const.zero_Celsius

# Examples
temp_c = 25
temp_f = 77

print(f"\nExample: {temp_c}°C = {celsius_to_kelvin(temp_c):.2f} K")
print(f"Example: {temp_f}°F = {fahrenheit_to_celsius(temp_f):.2f}°C")
print(f"Example: {temp_f}°F = {fahrenheit_to_kelvin(temp_f):.2f} K")

=== Temperature Constants ===
0°C in Kelvin: 273.15 K
1°F difference: 0.5555555555555556 K

Example: 25°C = 298.15 K
Example: 77°F = 25.00°C
Example: 77°F = 298.15 K


### Pressure Conversions
All pressure units convert **to Pascals (Pa)**:

```python
value_in_pascals = value_in_other_unit * conversion_factor
```

In [9]:
# Pressure conversions (to Pascals)
print("=== Pressure Conversions ===")
print("1 atmosphere =", const.atmosphere, "Pa")
print("1 bar =", const.bar, "Pa")
print("1 torr (mmHg) =", const.torr, "Pa")
print("1 mmHg =", const.mmHg, "Pa")
print("1 psi =", const.psi, "Pa")

# Example: Convert 2 atmospheres to bar
atm = 2
pascals = atm * const.atmosphere
bar = pascals / const.bar
print(f"\nExample: {atm} atm = {bar:.4f} bar")

=== Pressure Conversions ===
1 atmosphere = 101325.0 Pa
1 bar = 100000.0 Pa
1 torr (mmHg) = 133.32236842105263 Pa
1 mmHg = 133.32236842105263 Pa
1 psi = 6894.757293168361 Pa

Example: 2 atm = 2.0265 bar


### Energy Conversions
All energy units convert **to Joules (J)**:

```python
value_in_joules = value_in_other_unit * conversion_factor
```

In [10]:
# Energy conversions (to Joules)
print("=== Energy Conversions ===")
print("1 electron volt (eV) =", const.eV, "J")
print("1 calorie =", const.calorie, "J")
print("1 calorie (thermochemical) =", const.calorie_th, "J")
print("1 erg =", const.erg, "J")
print("1 British thermal unit =", const.Btu, "J")
print("1 ton of TNT =", const.ton_TNT, "J")

# Example: Convert 100 calories to kJ
calories = 100
joules = calories * const.calorie
kilojoules = joules / 1000
print(f"\nExample: {calories} cal = {kilojoules:.2f} kJ")

=== Energy Conversions ===
1 electron volt (eV) = 1.602176634e-19 J
1 calorie = 4.184 J
1 calorie (thermochemical) = 4.184 J
1 erg = 1e-07 J
1 British thermal unit = 1055.05585262 J
1 ton of TNT = 4184000000.0 J

Example: 100 cal = 0.42 kJ


### Power & Other Conversions

Power converts **to Watts (W)**:  
Other units have their base conversions:

In [11]:
# Power conversions (to Watts)
print("=== Power Conversions ===")
print("1 horsepower =", const.hp, "W")

# Area conversions (to square meters)
print("\n=== Area Conversions ===")
print("1 hectare =", const.hectare, "m²")
print("1 acre =", const.acre, "m²")

# Volume conversions (to cubic meters)
print("\n=== Volume Conversions ===")
print("1 liter =", const.liter, "m³")
print("1 gallon (US) =", const.gallon, "m³")
print("1 gallon (UK) =", const.gallon_imp, "m³")
print("1 fluid ounce (US) =", const.fluid_ounce, "m³")
print("1 fluid ounce (UK) =", const.fluid_ounce_imp, "m³")
print("1 barrel =", const.barrel, "m³")

# Speed conversions (to meters per second)
print("\n=== Speed Conversions ===")
print("1 km/h =", const.kmh, "m/s")
print("1 mph =", const.mph, "m/s")
print("1 knot =", const.knot, "m/s")

=== Power Conversions ===
1 horsepower = 745.6998715822701 W

=== Area Conversions ===
1 hectare = 10000.0 m²
1 acre = 4046.8564223999992 m²

=== Volume Conversions ===
1 liter = 0.001 m³
1 gallon (US) = 0.0037854117839999997 m³
1 gallon (UK) = 0.00454609 m³
1 fluid ounce (US) = 2.9573529562499998e-05 m³
1 fluid ounce (UK) = 2.84130625e-05 m³
1 barrel = 0.15898729492799998 m³

=== Speed Conversions ===
1 km/h = 0.2777777777777778 m/s
1 mph = 0.44703999999999994 m/s
1 knot = 0.5144444444444445 m/s


### Angle Conversions
All angle units convert **to radians**:

```python
value_in_radians = value_in_other_unit * conversion_factor
```

In [12]:
# Angle conversions (to radians)
print("=== Angle Conversions ===")
print("1 degree =", const.degree, "radians")
print("1 arcminute =", const.arcmin, "radians")
print("1 arcsecond =", const.arcsec, "radians")

# Verify conversions
print("\nVerification:")
print("180° in radians:", 180 * const.degree)
print("π (expected):", np.pi)
print("Match:", np.isclose(180 * const.degree, np.pi))

# Example: Convert 45 degrees to radians
degrees = 45
radians = degrees * const.degree
print(f"\nExample: {degrees}° = {radians:.4f} radians")
print(f"Expected: π/4 = {np.pi/4:.4f} radians")

=== Angle Conversions ===
1 degree = 0.017453292519943295 radians
1 arcminute = 0.0002908882086657216 radians
1 arcsecond = 4.84813681109536e-06 radians

Verification:
180° in radians: 3.141592653589793
π (expected): 3.141592653589793
Match: True

Example: 45° = 0.7854 radians
Expected: π/4 = 0.7854 radians


## Finding Constants - `find()` Function

Search for physical constants by keyword:

```python
scipy.constants.find(sub, disp=False)
```

- **sub**: Search string (case-insensitive)
- **disp**: If True, prints results; if False, returns list
- Returns list of constant names matching the search

In [13]:
# Find constants related to 'electron'
print("=== Constants containing 'electron' ===")
electron_constants = const.find('electron')
for name in electron_constants:
    print(f"  - {name}")

# Find constants related to 'light'
print("\n=== Constants containing 'light' ===")
light_constants = const.find('light')
for name in light_constants:
    print(f"  - {name}")

# Display results directly
print("\n=== Constants containing 'mass' (displayed) ===")
const.find('mass', disp=True)

=== Constants containing 'electron' ===
  - alpha particle-electron mass ratio
  - atomic mass unit-electron volt relationship
  - classical electron radius
  - deuteron-electron mag. mom. ratio
  - deuteron-electron mass ratio
  - electron charge to mass quotient
  - electron g factor
  - electron gyromag. ratio
  - electron gyromag. ratio in MHz/T
  - electron mag. mom.
  - electron mag. mom. anomaly
  - electron mag. mom. to Bohr magneton ratio
  - electron mag. mom. to nuclear magneton ratio
  - electron mass
  - electron mass energy equivalent
  - electron mass energy equivalent in MeV
  - electron mass in u
  - electron molar mass
  - electron relative atomic mass
  - electron to alpha particle mass ratio
  - electron to shielded helion mag. mom. ratio
  - electron to shielded proton mag. mom. ratio
  - electron volt
  - electron volt-atomic mass unit relationship
  - electron volt-hartree relationship
  - electron volt-hertz relationship
  - electron volt-inverse meter relations

## Physical Constants Database - `physical_constants`

Access detailed information about physical constants:

```python
scipy.constants.physical_constants[key]
```

Returns tuple: `(value, unit, uncertainty)`

- **value**: Numeric value of the constant
- **unit**: Unit string
- **uncertainty**: Standard uncertainty

In [14]:
# Access detailed constant information
speed_of_light = const.physical_constants['speed of light in vacuum']
print("Speed of light in vacuum:")
print(f"  Value: {speed_of_light[0]}")
print(f"  Unit: {speed_of_light[1]}")
print(f"  Uncertainty: {speed_of_light[2]}")

# Planck constant
planck = const.physical_constants['Planck constant']
print("\nPlanck constant:")
print(f"  Value: {planck[0]}")
print(f"  Unit: {planck[1]}")
print(f"  Uncertainty: {planck[2]}")

# Electron mass
electron_mass = const.physical_constants['electron mass']
print("\nElectron mass:")
print(f"  Value: {electron_mass[0]}")
print(f"  Unit: {electron_mass[1]}")
print(f"  Uncertainty: {electron_mass[2]}")

Speed of light in vacuum:
  Value: 299792458.0
  Unit: m s^-1
  Uncertainty: 0.0

Planck constant:
  Value: 6.62607015e-34
  Unit: J Hz^-1
  Uncertainty: 0.0

Electron mass:
  Value: 9.1093837139e-31
  Unit: kg
  Uncertainty: 2.8e-40


In [15]:
# List first 10 constants in the database
print("First 10 physical constants in database:")
for i, key in enumerate(list(const.physical_constants.keys())[:10]):
    print(f"{i+1}. {key}")

# Total number of constants
total = len(const.physical_constants)
print(f"\nTotal constants in database: {total}")

First 10 physical constants in database:
1. Wien displacement law constant
2. atomic unit of 1st hyperpolarizablity
3. atomic unit of 2nd hyperpolarizablity
4. atomic unit of electric dipole moment
5. atomic unit of electric polarizablity
6. atomic unit of electric quadrupole moment
7. atomic unit of magn. dipole moment
8. atomic unit of magn. flux density
9. deuteron magn. moment
10. deuteron magn. moment to Bohr magneton ratio

Total constants in database: 445


## Practical Example: Energy Calculations

Calculate the energy of a photon with wavelength λ:

\[ E = \frac{hc}{\lambda} \]

Where:
- \( h \) = Planck constant
- \( c \) = Speed of light
- \( \lambda \) = Wavelength

In [16]:
# Calculate energy of visible light photon
wavelength_nm = 550  # Green light in nanometers
wavelength_m = wavelength_nm * 1e-9  # Convert to meters

# Energy in Joules: E = hc/λ
energy_j = (const.h * const.c) / wavelength_m
print(f"Wavelength: {wavelength_nm} nm")
print(f"Energy: {energy_j:.4e} J")

# Convert to electron volts
energy_ev = energy_j / const.eV
print(f"Energy: {energy_ev:.2f} eV")

# Verify with formula: E(eV) ≈ 1240/λ(nm)
energy_approx = 1240 / wavelength_nm
print(f"\nApproximate formula: {energy_approx:.2f} eV")
print(f"Difference: {abs(energy_ev - energy_approx):.4f} eV")

Wavelength: 550 nm
Energy: 3.6117e-19 J
Energy: 2.25 eV

Approximate formula: 2.25 eV
Difference: 0.0003 eV


## Practical Example: Multi-Step Unit Conversion

Convert car fuel efficiency from miles per gallon (mpg) to liters per 100 km:

In [17]:
# US car fuel efficiency
mpg = 30  # miles per gallon

# Step 1: Convert miles to km
km_per_gallon = (mpg * const.mile) / 1000
print(f"Step 1: {mpg} mpg = {km_per_gallon:.2f} km/gallon")

# Step 2: Convert gallons to liters
km_per_liter = km_per_gallon / (const.gallon * 1000)
print(f"Step 2: = {km_per_liter:.2f} km/L")

# Step 3: Invert to L/100km (European standard)
liters_per_100km = 100 / km_per_liter
print(f"Step 3: = {liters_per_100km:.2f} L/100km")

print(f"\nFinal: {mpg} mpg = {liters_per_100km:.2f} L/100km")

Step 1: 30 mpg = 48.28 km/gallon
Step 2: = 12.75 km/L
Step 3: = 7.84 L/100km

Final: 30 mpg = 7.84 L/100km


## Common Use Cases

### 1. **Physics Calculations**
- Quantum mechanics (h, ℏ, c)
- Relativity (c, G)
- Thermodynamics (k, R, σ)

### 2. **Chemistry**
- Avogadro constant (N_A)
- Gas constant (R)
- Atomic masses

### 3. **Engineering**
- Unit conversions
- Energy calculations
- Pressure and temperature

### 4. **Astronomy**
- Astronomical unit
- Light year
- Parsec
- Gravitational constant

In [18]:
def convert_units(value, from_unit, to_unit, unit_type):
    """
    Convert between units using scipy.constants
    
    Parameters:
    - value: numeric value to convert
    - from_unit: source unit (e.g., 'mile', 'pound')
    - to_unit: target unit (e.g., 'meter', 'kg')
    - unit_type: 'distance', 'mass', 'time', etc.
    """
    # Convert to base unit
    if hasattr(const, from_unit):
        base_value = value * getattr(const, from_unit)
    else:
        raise ValueError(f"Unknown unit: {from_unit}")
    
    # Convert from base to target
    if to_unit == 'base':
        return base_value
    elif hasattr(const, to_unit):
        return base_value / getattr(const, to_unit)
    else:
        raise ValueError(f"Unknown unit: {to_unit}")

# Examples
print("5 miles to meters:", convert_units(5, 'mile', 'base', 'distance'))
print("100 pounds to kg:", convert_units(100, 'pound', 'base', 'mass'))
print("2 hours to seconds:", convert_units(2, 'hour', 'base', 'time'))

5 miles to meters: 8046.719999999999
100 pounds to kg: 45.35923699999999
2 hours to seconds: 7200.0


## Quick Reference Summary

### Most Used Constants:

| Category | Constant | Variable | Value (approx) |
|----------|----------|----------|----------------|
| Speed | Speed of light | `c` | 3.0 × 10⁸ m/s |
| Quantum | Planck constant | `h` | 6.626 × 10⁻³⁴ J·s |
| Gravity | Gravitational const | `G` | 6.674 × 10⁻¹¹ m³/(kg·s²) |
| Charge | Elementary charge | `e` | 1.602 × 10⁻¹⁹ C |
| Thermal | Boltzmann const | `k` | 1.381 × 10⁻²³ J/K |
| Chemistry | Avogadro constant | `N_A` | 6.022 × 10²³ 1/mol |
| Gas | Gas constant | `R` | 8.314 J/(mol·K) |
| Math | Pi | `pi` | 3.14159... |

### Quick Conversions:
- Distance: `mile`, `foot`, `inch`, `km` (→ meters)
- Mass: `pound`, `ounce`, `gram` (→ kg)
- Time: `hour`, `minute`, `day` (→ seconds)
- Energy: `eV`, `calorie`, `Btu` (→ Joules)
- Pressure: `atm`, `bar`, `psi` (→ Pascals)

## Summary: Key Takeaways

✓ **Import**: `import scipy.constants as const`  
✓ **Physical constants**: `const.c`, `const.h`, `const.G`, `const.k`  
✓ **Unit conversions**: Multiply by conversion factor to get base unit  
✓ **Search constants**: `const.find('keyword')`  
✓ **Detailed info**: `const.physical_constants[key]` returns (value, unit, uncertainty)  
✓ **CODATA values**: All constants from authoritative scientific sources  
✓ **Precision**: High-precision values with uncertainties included  

### Best Practices:
1. Always use constants instead of hardcoding values
2. Check units carefully when converting
3. Use `physical_constants` for uncertainty information
4. Remember temperature conversions are non-linear
5. Verify conversions with known relationships

### Next Steps:
- Explore `scipy.linalg` for linear algebra
- Learn `scipy.integrate` for numerical integration
- Study `scipy.optimize` for optimization problems